File: DriftGame/libs/src/java-universal-tween-engine/tween-engine-api/src/aurelienribon/tweenengine/equations/Quad.java
package aurelienribon.tweenengine.equations;

import aurelienribon.tweenengine.TweenEquation;

/**
 * Easing equation based on Robert Penner's work: http://robertpenner.com/easing/
 */
public abstract class Quad extends TweenEquation {
    public static final Quad IN = new Quad() {
        @Override
        public float compute(float t) {
            return t * t;
        }

        @Override
        public String toString() {
            return "Quad.IN";
        }
    };

    public static final Quad OUT = new Quad() {
        @Override
        public float compute(float t) {
            return -t * (t - 2);
        }

        @Override
        public String toString() {
            return "Quad.OUT";
        }
    };

    public static final Quad INOUT = new Quad() {
        @Override
        public float compute(float t) {
            if ((t *= 2) < 1) return 0.5f * t * t;
            return -0.5f * ((--t) * (t - 2) - 1);
        }

        @Override
        public String toString() {
            return "Quad.INOUT";
        }
    };
}

File: DriftGame/libs/src/java-universal-tween-engine/tween-engine-api/src/aurelienribon/tweenengine/equations/Quad.kt
package aurelienribon.tweenengine.equations

import aurelienribon.tweenengine.TweenEquation

/**
 * Easing equation based on Robert Penner's work: [Look here](http://robertpenner.com/easing/)
 */
abstract class Quad : TweenEquation() {
    companion object {
        val IN = object : Quad() {
            override fun compute(t: Float) = t * t
            override fun toString() = "Quad.IN"
        }

        val OUT = object : Quad() {
            override fun compute(t: Float) = -t * (t - 2)
            override fun toString() = "Quad.OUT"
        }

        val INOUT = object : Quad() {
            override fun compute(t: Float): Float {
                var t = t
                if ((2.let { t *= it; t }) < 1) return 0.5f * t * t
                return -0.5f * ((--t) * (t - 2) - 1)
            }

            override fun toString() = "Quad.INOUT"

        }
    }
}
